PerfFlowAspect C/C++ support

dongahn opened the pull request:
This PR is work in progress and adds C/C++ support for PerfFlowAspect.
I am posting this to get some early feedback. 

Unlike dynamic programming languages such as Python,
C/C++ cannot natively and easily weave PerfFlowAspect
into the target program (e.g., using language
features like Python decorator).
To overcome this language-level limitation,
PerfFlowAspect introduces a simple
[LLVM pass](https://llvm.org/docs/WritingAnLLVMPass.html#introduction-what-is-a-pass)
which can detect critical-path annotations
in the target C/C++ source files and to weave
these annotated program points with
PerfFlowAspect's runtime library as part
of Clang compilers' optimization pass.
This allows C/C++ and Python users alike to use the
consistent aspect-oriented programming (AOP) abstraction
of PerfFlowAspect.

Specifically, the PR:
- Create src directory and move perfflow python binding support in there
- Introduce c directory to add C/C++ support
- Vendor in libtap
- Use Flex and Bison to scan, parse and validate annotation string
- Add a shim layer that will contain an single API call
- Add TAP unit tests
- Add Weaving using LLVM Function Pass

Fix Issue #7, #8

@FrankD412: can you see the simple moving of python support into `src/python`
still looks good for your pip install support? 2750514

@ilagunap: can you give a quick review for the LLVM pass for this? e519d4c

@SteVwonder: I am tagging you as the overall reviewer. 

Before merging this, I still need to automate the running of test cases
using TAP and manually test it on Fluxion and fix any issues
if this doesn't work out of box before merging it.

dongahn commented:
I just applied this to Fluxion and this is working reasonable well! Still have a few tightening up is needed but nevertheless.

Source code change needed were only two 2 lines:
```diff
index 090c6bfd..0bea80a9 100644
--- a/resource/modules/resource_match.cpp
+++ b/resource/modules/resource_match.cpp
@@ -1766,6 +1766,7 @@ out:
     return rc;
 }

+__attribute__((annotate("@critical_path()")))
 static void match_request_cb (flux_t *h, flux_msg_handler_t *w,
                               const flux_msg_t *msg, void *arg)
 {

index 1c00d6d9..70c40a5c 100644
--- a/qmanager/modules/qmanager_callbacks.cpp
+++ b/qmanager/modules/qmanager_callbacks.cpp
@@ -394,6 +394,7 @@ int qmanager_safe_cb_t::jobmanager_hello_cb (flux_t *h, const flux_msg_t *msg,
     return rc;
 }

+__attribute__((annotate ("@critical_path()")))
 void qmanager_safe_cb_t::jobmanager_alloc_cb (flux_t *h, const flux_msg_t *msg,
                                               void *arg)
 {
```

Then for testing:

```console
$ cat input.sh
#! /bin/bash

flux mini submit -N 2 -n 2 hostname
flux mini submit -N 2 -n 2 hostname
flux mini submit -N 2 -n 2 hostname
flux mini submit -N 2 -n 2 hostname
flux mini submit -N 2 -n 2 hostname
flux mini submit -N 2 -n 2 hostname

flux queue drain

$ flux start -s 2 < input.sh
ƒqvbJP1
ƒyN1hPM
ƒ26PiHtB
ƒ2DhikUo
ƒ2L4sfef
ƒ2STWa6s
```

This generated a tracing file (note three tracing files are generated with two empty because of fork/exec so I need to work on this):

```console
$ ll
total 16
-rwxr--r-- 1 ahn1 ahn1  248 Feb 28 23:13 input.sh
-rw-r--r-- 1 ahn1 ahn1 3674 Feb 28 23:31 perfflow.out.docker-desktop.51933
-rw-r--r-- 1 ahn1 ahn1    2 Feb 28 23:31 perfflow.out.docker-desktop.51976
-rw-r--r-- 1 ahn1 ahn1    2 Feb 28 23:31 perfflow.out.docker-desktop.51980
```

The overall global timeline display shows the general activities of Fluxion across its two modules and performance profiles in the bottom as well. (Note that there are a lot of white spaces between activities and as we as the AOP to different components we will fill in these spaces):

![Screen Shot 2021-02-28 at 3 33 14 PM](https://user-images.githubusercontent.com/862123/109437729-380b5e80-79db-11eb-8e7c-df3707ac7a8e.png)

I then zoomed in one of the five activities as shown below. This immediately tells that a part of the problem is no overlapping between `sched-fluxion-qmanager` and `sched-fluxion-resource`

![Screen Shot 2021-02-28 at 3 34 25 PM](https://user-images.githubusercontent.com/862123/109437769-76a11900-79db-11eb-96d2-9b213b65f674.png)

Good start!

dongahn commented:
BTW build system integration wouldn't be too painful either once all is said and done:

```diff
index 65560f5b..32885f28 100644
--- a/qmanager/modules/Makefile.am
+++ b/qmanager/modules/Makefile.am
@@ -47,6 +47,7 @@ sched_fluxion_qmanager_la_SOURCES = \
     $(top_srcdir)/src/common/c++wrappers/eh_wrapper.hpp

 sched_fluxion_qmanager_la_CXXFLAGS = \
+    -Wc,-Xclang,-load,-Xclang,/usr/src/PerfFlowAspect/src/c/weaver/build/weave/libWeavePass.so \
     $(AM_CXXFLAGS) \
     $(FLUX_CORE_CFLAGS) \
     $(FLUX_SCHEDUTIL_CFLAGS)
@@ -56,9 +57,11 @@ sched_fluxion_qmanager_la_LIBADD = \
     $(FLUX_CORE_LIBS) \
     $(FLUX_SCHEDUTIL_LIBS) \
     $(CZMQ_LIBS) \
+    -lperfflow_runtime \
     $(JANSSON_LIBS)

 sched_fluxion_qmanager_la_LDFLAGS = \
+    -L/usr/src/PerfFlowAspect/src/c/runtime \
     $(AM_LDFLAGS) \
     $(fluxmod_ldflags) -module


index 802ad66f..17fb2f32 100644
--- a/resource/modules/Makefile.am
+++ b/resource/modules/Makefile.am
@@ -19,6 +19,7 @@ fluxmod_LTLIBRARIES = sched-fluxion-resource.la
 sched_fluxion_resource_la_SOURCES = \
     resource_match.cpp
 sched_fluxion_resource_la_CXXFLAGS = \
+    -Wc,-Xclang,-load,-Xclang,/usr/src/PerfFlowAspect/src/c/weaver/build/weave/libWeavePass.so \
     $(AM_CXXFLAGS) \
     $(JOBSPEC_CFLAGS) \
     $(FLUX_HOSTLIST_CFLAGS) \
@@ -31,10 +32,12 @@ sched_fluxion_resource_la_LIBADD = \
     $(DL_LIBS) \
     $(HWLOC_LIBS) \
     $(UUID_LIBS) \
+    -lperfflow_runtime \
     $(JANSSON_LIBS) \
     $(CZMQ_LIBS) \
     $(BOOST_LDFLAGS)
 sched_fluxion_resource_la_LDFLAGS = \
+    -L/usr/src/PerfFlowAspect/src/c/runtime \
     $(AM_LDFLAGS) \
     $(fluxmod_ldflags) -module
```

SteVwonder commented on src/c/weaver/weave/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+    perfflow_weave.cpp
+)
+
+# Use C++11 to compile our pass (i.e., supply -std=c++11).
```
The other cmake file sets the `CMAKE_CXX_STANDARD` to `14`, but this comment mentions that these targets are supported in `11`.  Should we loosen the `CMAKE_CXX_STANDARD`?

dongahn replied:
Our pass is C++ 11 but the overall LLVM Pass header files requires C++14. I will see if we say stay at C++14 though.

SteVwonder commented on src/c/test/smoketest.cpp:
```diff
@@ -0,0 +1,50 @@
+#include <string>
+#include <unistd.h>
+
+__attribute__((annotate("@critical_path(pointcut=\"after\", scope=\"bar\")")))
```
Thoughts on also supporting single-quotes within the annotation?

```suggestion
__attribute__((annotate("@critical_path(pointcut='after', scope='bar')")))
```

The escaped double-quotes aren't a big deal, but I think the ability to also use an unescaped single-quote would be nice.  I assume it would just require tweaking the regex in the yak/bison for the label to something like:

```diff
- label              ["]([^"\\\n]|\\(.|\n))*["]
+ label              ["']([^"'\\\n]|\\(.|\n))*["']
```

dongahn replied:
Oh.... good idea!

dongahn replied:
I realize this will validate quote mismatches as well. (pointcut='after\") is valid. You can avoid this but it feels like we may just stick with single quotes as the only valid gramma...

SteVwonder reviewed (commented):
This is really cool @dongahn! I love that the code compiles cleanly with the attributes even when you don't have the WeavePass clang pass setup/compiled.  Some minor comments/thoughts below, but generally looks good.

dongahn commented on src/c/runtime/advice_chrome_tracing.cpp:
```diff
@@ -0,0 +1,256 @@
+
+advice_chrome_tracing_t::~advice_chrome_tracing_t ()
+{
+}
```
CTF doesn't require the closing bracket. But it is an excellent point to use the destructor to do some cleaning+optimization. I will add buffering (so that the trace will go memory until it hit a certain size) and as part of that I will flush the remaining buffer and close the output file in the destructor.

dongahn commented on src/c/runtime/advice_chrome_tracing.cpp:
```diff
@@ -0,0 +1,256 @@
+}
+
+int advice_chrome_tracing_t::create_event (json_t **o,
+                                           const char *mn, const char *fn)
```
It is "module name" which I am using for "category" at the moment. Will expand the name.

dongahn commented on src/c/runtime/advice_chrome_tracing.cpp:
```diff
@@ -0,0 +1,256 @@
+        json_decref (event);
+        return rc;
+    }
+    if ( (rc = encode_event (event, "b", scope, m_before_counter)) < 0) {
```
In order for this to be effective, I think we need to expose id to be set at runtime. I just don't know yet as to how to do this. As you are "decorating" a function, you can only specify static properties to the annotation. If you can directly call APIs, then this is easy. But this PR isn't just there yet.

dongahn commented on src/c/parser/lex.l:
```diff
@@ -0,0 +1,51 @@
+%%
+
+[ \t]   ; // ignore all whitespace
+"@perfflow.aspect.critical_path" { return T_CPA; }
```
Good catch!

SteVwonder commented on src/c/runtime/advice_chrome_tracing.cpp:
```diff
@@ -0,0 +1,256 @@
+    m_ofs << "[" << std::endl;
+}
+
+advice_chrome_tracing_t::~advice_chrome_tracing_t ()
```
Does this destructor need to be defined as virtual?  Since we have shared pointers of type `advice_base_t` in the code, the `advice_chrome_tracing_t` destructor will not be called on `delete`, right?

dongahn replied:
Yes we should. Thank you for catching. I'm reworking the constructor/destructor a bit anyways.

dongahn commented:
Moving @ilagunap's comment to here:

> I suggest to test it on a recent version of LLVM (10.0 or later).
> I suggest to add debug information to the call function that you create (in case someone compiles with -g).
> You register the pass as EP_EarlyAsPossible. Did you check that these instrumentations are not removed by later optimizations (e.g., when compiling with -O3)?

dongahn commented:
I also found that I will need "flow" support to catch issues with "overlapping" so I will add some flow support into annotation.

![Screen Shot 2021-03-02 at 4 50 35 PM](https://user-images.githubusercontent.com/862123/109735905-5fe3f900-7b78-11eb-950d-5cc1c34c66e3.png)

dongahn commented:
@ilagunap: 

> I suggest to test it on a recent version of LLVM (10.0 or later).
I am not there yet but I will definitely get there as part of reduction to practice. The code isn't there yet. Do you see a reason why the code wouldn't work (after you comment in the `FunctionCallee` change?)

dongahn commented:
@ilagunap:

> I suggest to add debug information to the call function that you create (in case someone compiles with -g).

Good suggestion. I'm not that familiar with LLVM API. Can you suggest an example code or documentation to add DWARF under -g?

FrankD412 reviewed (changes requested):
The setup file needs fixing to get this to retain its pip installability. I have a working set of changes, but can't check them in here. Two things are here:

- Move the `pyproject.toml.bak` and `setup.py` to the root of the repository.
- Make the following changes to `setup.py`
```
diff --git a/src/python/setup.py b/src/python/setup.py
index f8f4ddf..5709d2c 100644
--- a/src/python/setup.py
+++ b/src/python/setup.py
@@ -9,7 +9,7 @@
 ##############################################################
 
 from os import path
-from setuptools import setup, find_packages
+from setuptools import setup, find_namespace_packages
 
 
 def load_readme():
@@ -32,7 +32,10 @@ setup(
     version="0.0.1",
     author="Dong H. Ahn, Stephen Herbein, James Corbett, Francesco Di Natale",
     author_email="ahn1@llnl.gov, herbein1@llnl.gov, corbett8@llnl.gov, dinatale3@llnl.gov",
-    packages=find_packages(),
+    package_dir = {
+        '': 'src/python'
+        },
+    packages=['perfflowaspect'],
     entry_points={},
     install_requires=[],
     extras_require={},
```

FrankD412 commented:
Oh -- one more thing; the `pyproject.toml` equivalent:

```
diff --git a/src/python/pyproject.toml.bak b/src/python/pyproject.toml.bak
index 90ceac7..8717124 100644
--- a/src/python/pyproject.toml.bak
+++ b/src/python/pyproject.toml.bak
@@ -16,6 +16,10 @@ authors = [
     "James Corbett <corbett8@llnl.gov>",
     "Francesco Di Natale <dinatale3@llnl.gov>",
 ]
+packages = [
+    { include = "perfflowaspect", from = "src/python" },
+]
+
 
 [tool.poetry.dependencies]
 python = "^3.6"
```

dongahn commented:
> You register the pass as EP_EarlyAsPossible. Did you check that these instrumentations are not removed by later optimizations (e.g., when compiling with -O3)?

It seems okay. 
```console
$ clang++ -O3 -Xclang -load -Xclang ../weaver/build/weave/libWeavePass.so smoketest.cpp -o smoketest -L/usr/src/PerfFlowAspect/src/c/runtime -lperfflow_runtime
$ ls -l perfflow*
ls: cannot access 'perfflow*': No such file or directory
$ LD_LIBRARY_PATH=/usr/src/PerfFlowAspect/src/c/runtime ./smoketest
<CUT>

$ ls -l perfflow*
-rw-r--r-- 1 ahn1 ahn1 1666 Mar  3 01:09 perfflow.docker-desktop.4727.pfw
```

dongahn commented:
I just confirmed that this works pretty well for LC TOSS3 -- at least it does what I need at this point. So this would be good time to take a quick look at those new commits once again and land. We will surely find many issues, but I will address them as-need basis.

dongahn commented:
Well just found that d513c6521f236fbfbbb681c974954735c52023e0 commit message descrive "in" vs "out" semantics in opposite way. I will change it.

SteVwonder reviewed (commented):
Changes look good! Just one question below:

dongahn commented on src/c/test/smoketest3.cpp:
```diff
@@ -1,29 +0,0 @@
-#include <string>
-#include <unistd.h>
```
I should have commented. W/ the changes in smoktest.cpp, the coverage of this is essential same as that one. The tests are the area that need to be beefed up later anyway.

SteVwonder commented on src/python/perfflowaspect/advice_chrome.py:
```diff
@@ -34,7 +34,7 @@ class ChromeTracingAdvice:
     # TODO: especially PERFLOW_OPTIONS="log_file=my_name.log"
     # TODO: support for TOML config
-    fn = "perfflow.out" + os.uname()[1] + "." + str(os.getpid())
+    fn = "perfflow.out" + os.uname()[1] + "." + str(os.getpid()) + ".pfw"
```
```suggestion
    fn = "perfflow." + os.uname()[1] + "." + str(os.getpid()) + ".pfw"
```

dongahn commented:
@SteVwonder and @FrankD412: as discussed, I cleaned up the PR a bit and added a few new mods to this PR. Ready for a quick review before being merged. Once this goes in, PR #17 needs to be rebased at which point I will give my review. Thanks.

SteVwonder reviewed (approved):
LGTM!

dongahn commented:
Thanks @SteVwonder!
